project4/binomialengine.hpp

```diff
@@ -128,8 +129,29 @@ namespace QuantLib {
         // binomial tree 
         // (see J.C.Hull, "Options, Futures and other derivatives", 6th edition, pp 397/398)
 
+
         // Rollback to third-last step, and get underlying prices (s2) &
         // option values (p2) at this point
+
+        Real* prices_pen = new Real[timeSteps_ - 1];
+        Real* underlyings_pen = new Real[timeSteps_ - 1];
+
+        if (oscillations_ == true) {
+            option.rollback(grid[timeSteps_ - 2]);
+            Array vapenu(option.values());
+            QL_ENSURE(vapenu.size() == timeSteps_ - 1, "Expect time steps minus 1 nodes in grid at penultimate step");
+            for (int i = 0; i < timeSteps_-2; i++) {
+                underlyings_pen[i] = lattice->underlying(timeSteps_ - 2, i);
+                BlackCalculator bs_option(payoff->optionType(), payoff->strike(), underlyings_pen[i], v*std::sqrt(maturity/timeSteps_), std::exp(-r* maturity/ timeSteps_));
+                Real value_bs = bs_option.value();
+                prices_pen[i] = value_bs;
+                option.values()[i] = prices_pen[i];
```

Comment:
You can simplify this as
```suggestion
                option.values()[i] = bs_option.value();
```
so you don't need an array.